src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/CodeMetricsAnalyzer.cs

```diff
@@ -0,0 +1,485 @@
+
+                        default:
+                            throw new NotImplementedException();
+                    }
+                }
+
+                uint? getDefaultThreshold(string ruleId, SymbolKind symbolKind)
+                {
+                    if (!isApplicableByDefault(ruleId, symbolKind))
+                    {
+                        return null;
+                    }
+
+                    // Compat: we match the default threshold values for old FxCop implementation.
+                    switch (ruleId)
+                    {
+                        case CA1501RuleId:
+                            return 5;
+
+                        case CA1502RuleId:
+                            return 25;
+
+                        case CA1505RuleId:
+                            return 10;
+
+                        case CA1506RuleId:
+                            return symbolKind == SymbolKind.NamedType ? 95 : (uint)40;
+
+                        default:
+                            throw new NotImplementedException();
+                    }
+                }
+            });
+        }
+
+        private static bool TryGetRuleIdToThresholdMap(
+            ImmutableArray<AdditionalText> additionalFiles,
+            CancellationToken cancellationToken,
+            out AdditionalText additionalText,
+            out ImmutableDictionary<string, List<(SymbolKind?, uint)>> ruleIdToThresholdMap,
+            out List<Diagnostic> invalidFileDiagnostics)
+        {
+            invalidFileDiagnostics = null;
+            ruleIdToThresholdMap = null;
+
+            // Parse the additional file for code metrics configuration.
+            // Return false if there is no such additional file or it contains at least one invalid entry.
+            additionalText = TryGetCodeMetricsConfigurationFile(additionalFiles, cancellationToken);
+            return additionalText != null &&
+                TryParseCodeMetricsConfigurationFile(additionalText, cancellationToken, out ruleIdToThresholdMap, out invalidFileDiagnostics);
+        }
+
+        private static AdditionalText TryGetCodeMetricsConfigurationFile(ImmutableArray<AdditionalText> additionalFiles, CancellationToken cancellationToken)
+        {
+            StringComparer comparer = StringComparer.Ordinal;
+            foreach (AdditionalText textFile in additionalFiles)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                string fileName = Path.GetFileName(textFile.Path);
+                if (comparer.Equals(fileName, CodeMetricsConfigurationFile))
+                {
+                    return textFile;
+                }
+            }
+
+            return null;
+        }
+
+        private static bool TryParseCodeMetricsConfigurationFile(
+            AdditionalText additionalText,
+            CancellationToken cancellationToken,
+            out ImmutableDictionary<string, List<(SymbolKind?, uint)>> ruleIdToThresholdMap,
```

Comment:
reason it is List rather than dictionary? I guess it is expected to be small most of time?

Reply:
IReadOnlyList?

Reply:
Yes, only one rule realistically needs separate values for Types and Methods. I can change it to IReadOnlyList.